### DISCUSSION TITLE
Capsule Cast

### DISCUSSION THREAD

**Discussion by ViachaslauBusel** [bepu/bepuphysics2#183] [Q&A] [answered]
Hello, could you please explain how "Simulation.Sweep" works. On the example of a capsule.
`Simulation.Sweep(
                    new Capsule(radius, height),
                    new RigidPose(origin + Vector3.UnitY * (height / 2.0f), Quaternion.Identity),
                    new BodyVelocity(normalizedTargetDirection),
                    targetDistance,
                    map.BufferPool,
                    ref  handler);`
![Physics-OverlapCapsule2-small](https://user-images.githubusercontent.com/15468413/145863269-2858f490-3db7-4901-a468-1138ce67dfdb.jpg)

`  public void OnHit(ref float maximumT, float t, in Vector3 hitLocation, in Vector3 hitNormal, CollidableReference collidable)`
t - Is this distance?

here is an example of how it works in Unity
![capsulecast](https://user-images.githubusercontent.com/15468413/145864826-8c7ddba6-ca1f-4fcf-bd7c-25396412682b.png)
the red area is the collision detection area

**Comment by RossNordby** [comment, marked as answer]
`t` is the time to impact, given the input sweep velocity. For example, for a pure linear sweep, if the velocity has a magnitude of 2 and the t value is 0.5, then the distance traveled to the impact is 2 * 0.5 = 1.

The `t` value is used instead of distance because distance ceases to be the most useful output when a sweep has nonzero angular velocity. For ray casts or linear-only sweeps, the time of impact can also be used to multiply the ray direction or linear velocity to get displacement easily.

(As a side note, in bepuphysics2, capsules do not take a full 'height' in their constructor- instead, it takes the length of the internal line segment. A capsule with 0 length would just behave like a sphere.)
